### PR TITLE
LUCENE-9276: Use same code-path for updateDocuments and updateDocument

### PR DESCRIPTION
**Pull request by s1monw**
Today we have a large amount of duplicated code that is rather of
complex nature. This change consolidates the code-paths to always
use the updateDocuments path.

### CODE REVIEW

**Comment by s1monw** [comment]
@mikemccand @dnhatn wanna take a look

**Comment by s1monw** [comment]
@uschindler I would love to get your input here too

File: lucene/core/src/java/org/apache/lucene/index/DocumentsWriterPerThread.java [hunk 1]
```diff
@@ -346,7 +285,7 @@ public long updateDocuments(Iterable<? extends Iterable<? extends IndexableField
     }
   }
   
-  private long finishDocument(DocumentsWriterDeleteQueue.Node<?> deleteNode) {
+  private long finishDocument(DocumentsWriterDeleteQueue.Node<?> deleteNode, int docCount) {
```
Comment (dnhatn) [inline lucene/core/src/java/org/apache/lucene/index/DocumentsWriterPerThread.java]: Should we call this finishDocument**S** ?

File: lucene/core/src/java/org/apache/lucene/index/DocumentsWriter.java [hunk 2]
```diff
@@ -474,49 +474,11 @@ long updateDocuments(final Iterable<? extends Iterable<? extends IndexableField>
     return seqNo;
   }
 
+
   long updateDocument(final Iterable<? extends IndexableField> doc, final Analyzer analyzer,
```
Comment (dnhatn) [inline lucene/core/src/java/org/apache/lucene/index/DocumentsWriter.java]: Can we also remove this method and delegate `updateDocument` to `updateDocuments` in IndexWriter instead?

File: lucene/core/src/java/org/apache/lucene/index/IndexWriter.java [hunk 3]
```diff
@@ -1655,7 +1655,7 @@ private long updateDocument(final DocumentsWriterDeleteQueue.Node<?> delNode,
     ensureOpen();
     boolean success = false;
     try {
-      final long seqNo = maybeProcessEvents(docWriter.updateDocument(doc, analyzer, delNode));
+      final long seqNo = maybeProcessEvents(docWriter.updateDocuments(List.of(doc), analyzer, delNode));
```
Comment (dnhatn) [inline lucene/core/src/java/org/apache/lucene/index/IndexWriter.java]: Can we just delegate this method to `IndexWriter#updateDocuments`?

**Pull request merged**